use-cases/src/Contracts/CrowdFunding.hs

```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+-- This is the fully parallel version that collects all contributions
+-- in a single transaction. This is, of course, limited by the maximum
+-- number of inputs a transaction can have.
+
+module Contracts.CrowdFunding (contribute, collect, refund)
+where
+
+import           Control.Monad (void)
+import           Plutus
+{-# ANN module "HLint: ignore Avoid restricted function" #-}
+
+
+-- Functionality for campaign contributors
+-- --
+
+-- Contribute funds to the campaign (contributor)
+--
+contribute :: PubKey -> Value -> T TxOutRef
+contribute ownerPubKey value
+  = do
+    { assert (value > 0)
+    ; contributorPubKey <- lookupMyPubKey
+    ; myPayment         <- createPayment (value + standardTxFee)
+    ; submitTransaction Tx
+      { txInputs  = [myPayment]
+      , txOutputs = [TxOutScript
+                       value
+                       (contributionScript ownerPubKey)
+                       [| contributorPubKey |] -- data script: ought to be lifted into PLC at coordination runtime
+                    ]
+      }
+    }
+    -- the transaction above really ought to be merely a transaction *template* and the transaction fee ought to be
+    -- added by the Wallet API Plutus library on the basis of the size and other costs of the transaction
+
+contributionScript :: PubKey -> Validator
+contributionScript ownerPubKey
+  = -- needs a splice adding the function triggering the Plutus Tx compiler plugin (compiled at coordination compile time)
+    --
+    -- What data is available inside the validator script? 
+    -- 
+    -- The transaction references its input via their hashes, so we need to 
+    -- de-reference them to be able to look at the coin values of their
+    -- outpus. De-referencing of a hash should probably also cost money 
+    -- (otherwise we could jam the validator by asking for data all the way
+    -- back to the beginning of the chain)
+    --
+    -- (see https://github.com/input-output-hk/cardano-sl/blob/develop/docs/block-processing/types.md for a description of the data types that exist in a block)
+    -- 
+    [| \redeemer txn contributorPubKey ->
+      let inputFunds = sum $ (sum $ view outputs) <$> inputs txn
+          payToOwner = currentBlockHeight > paymentDeadline &&

```

Comment:
this stuff is all references to undefined things, but I assume you know that. Would be cleaner to declare them explicitly as undefined so we know what's missing.